zebra-network/src/peer_set/initialize/tests/vectors.rs

```diff
@@ -587,6 +593,367 @@ async fn crawler_peer_limit_default_connect_ok_stay_open() {
+    zebra_test::init();
+
+    // This test requires an IPv4 network stack with 127.0.0.1 as localhost.
+    if zebra_test::net::zebra_skip_network_tests() {
+        return;
+    }
+
+    let unreachable_inbound_handshaker = service_fn(|_| async {
+        unreachable!("inbound handshaker should never be called with a zero peer limit")
+    });
+
+    let (_config, mut peerset_tx) =
+        spawn_inbound_listener_with_peer_limit(0, unreachable_inbound_handshaker).await;
+
+    let peer_result = peerset_tx.try_next();
+    assert!(
+        // `Err(_)` means that no peers are available, and the sender has not been dropped.
+        // `Ok(None)` means that no peers are available, and the sender has been dropped.
+        matches!(peer_result, Err(_) | Ok(None)),
+        "unexpected peer when inbound limit is zero: {:?}",
+        peer_result,
+    );
+}
+
+/// Test the listener with an inbound peer limit of one peer, and a handshaker that always errors.
+#[tokio::test]
+async fn listener_peer_limit_one_handshake_error() {
+    zebra_test::init();
+
+    // This test requires an IPv4 network stack with 127.0.0.1 as localhost.
+    if zebra_test::net::zebra_skip_network_tests() {
+        return;
+    }
+
+    let error_inbound_handshaker =
+        service_fn(|_| async { Err("test inbound handshaker always returns errors".into()) });
+
+    let (_config, mut peerset_tx) =
+        spawn_inbound_listener_with_peer_limit(1, error_inbound_handshaker).await;
+
+    let peer_result = peerset_tx.try_next();
+    assert!(
+        // `Err(_)` means that no peers are available, and the sender has not been dropped.
+        // `Ok(None)` means that no peers are available, and the sender has been dropped.
+        matches!(peer_result, Err(_) | Ok(None)),
+        "unexpected peer when all handshakes error: {:?}",
+        peer_result,
+    );
+}
+
+/// Test the listener with an inbound peer limit of one peer,
+/// and a handshaker that returns success then disconnects the peer.
+#[tokio::test]
+async fn listener_peer_limit_one_handshake_ok_then_drop() {
+    zebra_test::init();
+
+    // This test requires an IPv4 network stack with 127.0.0.1 as localhost.
+    if zebra_test::net::zebra_skip_network_tests() {
+        return;
+    }
+
+    let success_disconnect_inbound_handshaker = service_fn(|req: HandshakeRequest| async move {
+        let HandshakeRequest {
+            tcp_stream,
+            connected_addr: _,
+            connection_tracker,
+        } = req;
+
+        let (server_tx, _server_rx) = mpsc::channel(0);
+        let (shutdown_tx, _shutdown_rx) = oneshot::channel();
+        let error_slot = ErrorSlot::default();
+
+        let fake_client = peer::Client {
+            shutdown_tx: Some(shutdown_tx),
+            server_tx,
+            error_slot,
+        };
+
+        // Actually close the connection.
+        std::mem::drop(connection_tracker);
+        std::mem::drop(tcp_stream);
```

Comment:
Minor: I think this could be simplified to remove some bindings:

```suggestion
        let (server_tx, _server_rx) = mpsc::channel(0);
        let (shutdown_tx, _shutdown_rx) = oneshot::channel();
        let error_slot = ErrorSlot::default();

        let fake_client = peer::Client {
            shutdown_tx: Some(shutdown_tx),
            server_tx,
            error_slot,
        };

        // Actually close the connection.
        std::mem::drop(req);
```

Reply:
I would prefer to explicitly drop the two fields that actually matter in the tests: the TCP connection, and the connection tracker.